File: ruby/find_first_duplicate.rb
# require "pry"

def find_first_duplicate(arr)
  # type your code in here
  obj = {}

  for elem in arr
    if obj.key?(elem)
      return elem
    else
      obj[elem] = true
      # binding.pry
    end
  end
  -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  // type your code here
  const obj = {}
  for (let i = 0; i < arr.length; i++) {
    const elem = arr[i];
    if (obj[elem]) {
      return elem
    }
    obj[elem] = true
  }
  return -1
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Please add your pseudocode to this file
// iterate through Array
// keep track/ store what numbers i have seen

// if this number has already been return it
// if i go through the loop without finding a dupe return -1 

// And a written explanation of your solution
// create empty object
// loop through Array
// at first index assign to object a key of index and a value of true
// at the next index 
//   check if it's value is a key of the object
//     true then return the key
//     false assign this item as the key of the object with a value of true.valueOf

